### ISSUE TITLE
Cross-references broken?

### ISSUE THREAD

**Issue by nmueller18** [fiduswriter/fiduswriter#1176]
It seems that at least in FW 3.10.6, cross-referencing is broken. It is possible to add a cross-reference, and the respective text of the cross-reference (e. g. chapter heading) is also added. However, this text is not linked to the actual cross-reference, at least not in the same way as internal links are.
Or am I missing something obvious?

**Comment by johanneswilm** [comment]
As far as I can tell, it is working as it always did - it's mentioning the text and not linking to the source of origin. If you export to HTML though, it is also adding a link.

Are you requesting that it should add a link not only in the exported file but also while in the editor, @nmueller18 ?

**Comment by johanneswilm** [comment]
@nmueller18 Even if you use an internal link, you cannot just click that link and be transported to the anchor while in the editor. The reason for this is that it would create a bad experience for users if they accidentally click somewhere in the document and that then causes the editor to scroll.

**Comment by nmueller18** [comment]
> Even if you use an internal link, you cannot just click that link and be transported to the anchor while in the editor. The reason for this is that it would create a bad experience for users if they accidentally click somewhere in the document and that then causes the editor to scroll.

I agree with that. I think the behaviour of internal links is striking a good balance in terms of useability: you do not risk to be transported somewhere without consciously clicking twice but you have the opportunity to do so if you want.
To have this functionality also in FW (and not only in the exported document) is useful, IMHO, if you want to check a figure, for example. Even more useful, of course, would be the option to be taken back to the location before clicking the link …

**Comment by nmueller18** [comment]
> If you export to HTML though, it is also adding a link.
A heretic question: What is then the point to differentiate between internal links and cross references at all? I have tried to come up with a use-case where the differentiation between links and cross references would make sense but was not successful. Or is this already part of the implementation of creating cross references between documents?

BTW, I noticed that you can change an Internal link to Cross reference but not vice-versa.

**Comment by johanneswilm** [comment]
> What is then the point to differentiate between internal links and cross references at all?

They are two different kinds of animals: 

- In the case oof cross references, the user does not get to decide the linking text. It is a fixed text such as "Figure 1" or so. If you export to a cross reference aware format, you can make this appear as something else, for example "Figure 1, page 15". It's just when you export it to HTML, which has no page numbers, etc. that it gets exported as a link. If Figure 1 turns into Figure 2, the system will also change the text used in all your cross references to Figure 1. 
- Internal links, on the other hand, are a type of styling on some content. If you have nothing selected at the time the internal link is created, the dialog will he
lpyou by inserting a piece of text that corresponds to the thing you are linking to - for example "Figure 1". But you can later change it and the system will not attempt to change it for you. So if Figure 1 turns into Figure 2, the text will not be changed".

> BTW, I noticed that you can change an Internal link to Cross reference but not vice-versa.

Oh yeah I noticed the UI is inconsistent.

**Comment by johanneswilm** [comment]
> I have tried to come up with a use-case where the differentiation between links and cross references would make sense but was not successful. 

I think in many cases users will either use one or the other throughout a document. If you want to print the document or you want the system to handle the linking text for you, you will probably use cross references. If, on the other hand, your final product is mainly for the web or a PDF for on-screen reading, then you will probably be more interested in internal links.

For a long time we thought that it would be enough to provide internal links because the two seem somewhat similar, but we have had to accept that that is not really the case. The two behave differently and while few will be using both, it's impossible to make one and the same behave the way the internal links community wants and the way the cross reference community wants to at the same time. Some want the system to update the linking text - others do not want that to happen. etc.

> Or is this already part of the implementation of creating cross references between documents?

**Comment by johanneswilm** [comment]
> Or is this already part of the implementation of creating cross references between documents?

No, no work in that rea has been done so far as far as I am aware.

**Comment by johanneswilm** [comment]
> I think the behaviour of internal links is striking a good balance in terms of useability: you do not risk to be transported somewhere without consciously clicking twice but you have the opportunity to do so if you want.
> To have this functionality also in FW (and not only in the exported document) is useful, IMHO, if you want to check a figure, for example.

Ok, I guess we could offer that in the top-up dialog the same way it works for internal links.  

>  Even more useful, of course, would be the option to be taken back to the location before clicking the link …

I don't think I understand what you are asking for here.

**Comment by nmueller18** [comment]
> For a long time we thought that it would be enough to provide internal links because the two seem somewhat similar, but we have had to accept that that is not really the case. The two behave differently and while few will be using both, it's impossible to make one and the same behave the way the internal links community wants and the way the cross reference community wants to at the same time. Some want the system to update the linking text - others do not want that to happen. etc.

Thank you for explaining the reasoning behind Internal links and Cross references in detail! I do understand it better now. I am mainly thinking from the point of interactive pdfs where internal links make more sense. Perhaps some kind of explaining text directly in the window, i. e. available via a help-icon, would be useful to explain the differences in more detail?
I agree with you that users will probably want to use either "Internal links" or "Cross references" but not both. This duality, however, is clouded by the layout of the window, which presents "Cross references", "Internal" and "External" as three equivalent options. Perhaps it might be an idea to set "External" a little bit apart from the other two?

**Comment by nmueller18** [comment]
> Ok, I guess we could offer that in the top-up dialog the same way it works for internal links.

Hm, yes and no. After the discussion above, it might also let to confusion. The user might then expect links in their output but get plain text.

**Comment by nmueller18** [comment]
> > Even more useful, of course, would be the option to be taken back to the location before clicking the link …

> I don't think I understand what you are asking for here.

In pdfs, you can easily get back where you have been before clicking a link. In FW this is not possible (at least not consistently, perhaps also depending on OS and browser). Obviously, following the internal link is not stored in the browser history. Instead, clicking the "Back"-button brings me back to the document overview. [Here](https://baymard.com/blog/back-button-expectations) is an interesting discussion about problems with user expectations with the "Back"-button. Could `history.pushState()` be a solution?

**Comment by johanneswilm** [comment]
> Could `history.pushState()` be a solution?

When you move between the overview page and the editor and other "pages", you don't actually navigate to another page as that would make it much slower. We just use `history.pushState()` to emulate it. So technically, we could also use it here.

But it is tricky. I would think that if the user clicks the back-button in the editor, they probably want to get back to the document overview page and will be surprised if instead they have to click it for example 15 times before they actually get out. As the research you linked to suggests, users generally want the back button to mean "back to the last page".

**Comment by johanneswilm** [comment]
> Hm, yes and no. After the discussion above, it might also let to confusion. The user might then expect links in their output but get plain text.

The default output will be text that is also a link. The link won't be noticeable once you print it out, obviously. So I think it should be ok to add the link in the editor as well.

**Comment by nmueller18** [comment]
> But it is tricky. I would think that if the user clicks the back-button in the editor, they probably want to get back to the document overview page and will be surprised if instead they have to click it for example 15 times before they actually get out. As the research you linked to suggests, users generally want the back button to mean "back to the last page".

To be honest, in this respect I found the behaviour of FW always irritating. Generally, I use the "Back"-button as seldom as possible because with some pages you loose your input when you do. I would not have expected to be brought to the document overview.
I think the current behaviour is also counter-intuitive because in many scenarios you will open a document via an external link. Then, the "Back"-button as shortcut for getting to the document overview figuratively makes little sense. For that, you have the menu-entry or, alternatively, the close-button (which could perhaps be a little bit more prominent) just as you would have in a desktop-texteditor.
Therefore, if it was my decision, I would actually limit the functionality of the "Back"-button to jumps within the open document.

**Comment by johanneswilm** [comment]
> I use the "Back"-button as seldom as possible because with some pages you loose your input when you do.

This we probably should address. Which pages is this on? We should probably show a warning about losing information if one clicks the back button after having filled out half a form. On the editor page everything should be saved, sop it shouldn't be a problem there.

> I think the current behaviour is also counter-intuitive because in many scenarios you will open a document via an external link. Then, the "Back"-button as shortcut for getting to the document overview figuratively makes little sense.

As far as I can tell, if I just enter an external link and then click the "back" button, I am transported to the previous page and not the document overview page.

> Therefore, if it was my decision, I would actually limit the functionality of the "Back"-button to jumps within the open document.

I just looked at how Google Docs does it. They open a new tab when you open a document, so there is no "back to overview page" functionality there. But also if you create an internal link and jump to it, the back button doesn't work to jump back. I think we should probably collect a little more data on how people expect it to work. The link you sent seemed to say that it should bring people back to another page, but maybe people see it differently in a document editor environment. I wonder if this could be part of a small survey of some kind.

**Comment by nmueller18** [comment]
> This we probably should address. Which pages is this on? We should probably show a warning about losing information if one clicks the back button after having filled out half a form. On the editor page everything should be saved, sop it shouldn't be a problem there.

Sorry for the confusion. I meant other pages in the WWW.

> As far as I can tell, if I just enter an external link and then click the "back" button, I am transported to the previous page and not the document overview page.

I had a situation in mind like receiving a share-invitation via e-mail. I think this will transport you directly to the document.

> I think we should probably collect a little more data on how people expect it to work. The link you sent seemed to say that it should bring people back to another page, but maybe people see it differently in a document editor environment. I wonder if this could be part of a small survey of some kind.

This sounds like an excellent idea!
